components/support/migration/src/main/java/mozilla/components/support/migration/AbstractMigrationService.kt

```diff
@@ -0,0 +1,131 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.support.migration
+
+import android.app.NotificationChannel
+import android.app.NotificationManager
+import android.app.Service
+import android.content.Context
+import android.content.Intent
+import android.os.Build
+import android.os.IBinder
+import androidx.core.app.NotificationCompat
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+import mozilla.components.support.base.ids.NotificationIds
+import mozilla.components.support.base.log.logger.Logger
+
+private const val NOTIFICATION_TAG = "mozac.support.migration.notification"
+private const val NOTIFICATION_CHANNEL_ID = "mozac.support.migration.generic"
+
+// Non-final strings and icon.
+// https://github.com/mozilla-mobile/android-components/issues/5141
+private const val TEMPORARY_NOTIFICATION_TITLE = "Migrating"
+private const val TEMPORARY_NOTIFICATION_TEXT = "Migrating user data."
+private const val TEMPORARY_NOTIFICATION_CHANNEL_NAME = "Migration"
+private val temporaryNotificationIcon = R.drawable.mozac_lib_crash_notification
+
+/**
+ * Abstract implementation of a background service running a configured [FennecMigrator].
+ *
+ * An application using this implementation needs to extend this class and provide a
+ * [FennecMigrator] instance.
+ */
+abstract class AbstractMigrationService : Service() {
+    protected abstract val migrator: FennecMigrator
+
+    private val logger = Logger("MigrationService")
+    private val scope = MainScope()
+
+    override fun onCreate() {
+        super.onCreate()
+
+        logger.debug("Migration service started")
+
+        showForegroundNotification()
+
+        scope.launch {
+            try {
+                migrate()
+            } finally {
+                logger.debug("Stopping migration service")
+                shutdown()
+            }
+        }
+    }
+
+    override fun onBind(intent: Intent?): IBinder? = null
+
+    private suspend fun migrate() {
+        val results = migrator.migrateAsync().await()
```

Comment:
We'll need to emit a glean ping out of this; is that something we can do in a service? or would we need to communicate w/ the host app? It'd be great to not have to worry about making our results Parcelable...

But I guess we can deal with this later.

Reply:
That should just work. We are already using Glean from some other services (e.g. media).